crates/ruff_linter/src/rules/refurb/rules/verbose_decimal_constructor.rs

```diff
@@ -75,22 +76,58 @@ pub(crate) fn verbose_decimal_constructor(checker: &mut Checker, call: &ast::Exp
             value: str_literal, ..
         }) => {
             // Parse the inner string as an integer.
-            let trimmed = str_literal.to_str().trim_whitespace();
-
+            //
+            // For reference, a string argument to `Decimal` is parsed in CPython
+            // using this regex:
+            // https://github.com/python/cpython/blob/ac556a2ad1213b8bb81372fe6fb762f5fcb076de/Lib/_pydecimal.py#L6060-L6077
+            // _after_ trimming whitespace from the string and removing all occurrences of "_".
+            let mut trimmed = Cow::from(str_literal.to_str().trim_whitespace());
+            if memchr::memchr(b'_', trimmed.as_bytes()).is_some() {
+                trimmed = Cow::from(trimmed.replace('_', ""));
+            }
             // Extract the unary sign, if any.
-            let (unary, rest) = if let Some(trimmed) = trimmed.strip_prefix('+') {
-                ("+", trimmed)
+            let (unary, mut rest) = if let Some(trimmed) = trimmed.strip_prefix('+') {
+                ("+", Cow::from(trimmed))
             } else if let Some(trimmed) = trimmed.strip_prefix('-') {
-                ("-", trimmed)
+                ("-", Cow::from(trimmed))
             } else {
                 ("", trimmed)
             };
+            // Extract the exponent, if any.
+            let mut e_indices = memchr::memrchr2_iter(b'e', b'E', rest.as_bytes());
+            if let Some(index) = e_indices.next() {
+                // More than one `e` is an error of type `decimal.InvalidOperation`
+                // when calling Decimal on a string. The suggested
+                // fix would turn this into a SyntaxError. To maintain
+                // the behavior of the code, we abort the check in this case.
+                if e_indices.next().is_some() {
+                    return;
+                }
+                // This range will not cause a panic: in the worst case,
+                // 'e' was the last character in `rest`, and then the right
+                // hand side will be `""`.
+                let exponent = rest[index + 1..]
+                    .strip_prefix('+')
+                    .unwrap_or(&rest[index + 1..]);
+                // Verify that the exponent is a nonnegative integer
+                if !exponent.bytes().all(|c| c.is_ascii_digit()) {
+                    return;
+                };
+                // NB: We need not convert, e.g., `2e3` to `2000`.
+                // However, `2e+3` is a syntax error, so we remove
+                // the sign on the exponent.
+                rest = Cow::from(rest.replace('+', ""));
```

Comment:
`replace` is technically `O(n)`. Not that it matters much here but I think I would find it slightly more readable if we explicitly tested for the `+` sign and the expected position.

```
if rest[index + 1] == b'+' {
    rest.into_owned().remove(index + 1)
}
```

The alternative is to use `remove_matches` which captures the semantic better than `repalce` with a empty string

Reply:
I like `remove_matches` but isn't it nightly only? https://doc.rust-lang.org/std/string/struct.String.html#method.remove_matches

Reply:
Ah dang. I guess replace is still the best option